src/System.Private.CoreLib/src/System/Runtime/Loader/ComponentDependencyResolver.cs

```diff
@@ -0,0 +1,282 @@
+﻿using System.Collections.Generic;
+using System.IO;
+using System.Reflection;
+using System.Runtime.InteropServices;
+using Internal.IO;
+
+namespace System.Runtime.Loader
+{
+    public sealed class ComponentDependencyResolver
+    {
+        /// <summary>
+        /// The name of the neutral culture (same value as in Variables::Init in CoreCLR)
+        /// </summary>
+        private const string NeutralCultureName = "neutral";
+
+        /// <summary>
+        /// The extension of resource assembly (same as in BindSatelliteResourceByResourceRoots in CoreCLR)
+        /// </summary>
+        private const string ResourceAssemblyExtension = ".dll";
+
+        private readonly string _componentAssemblyPath;
+        private readonly Dictionary<string, string> _assemblyPaths;
+        private readonly string[] _nativeSearchPaths;
+        private readonly string[] _resourceSearchPaths;
+        private readonly string[] _componentDirectorySearchPaths;
+
+        public ComponentDependencyResolver(string componentAssemblyPath)
+        {
+            _componentAssemblyPath = componentAssemblyPath;
+
+            string assemblyPathsList = null;
+            string nativeSearchPathsList = null;
+            string resourceSearchPathsList = null;
+            int returnCode = 0;
+
+            try
+            {
+                returnCode = corehost_resolve_component_dependencies(
+                    componentAssemblyPath,
+                    (assembly_paths, native_search_paths, resource_search_paths) =>
+                    {
+                        assemblyPathsList = assembly_paths;
+                        nativeSearchPathsList = native_search_paths;
+                        resourceSearchPathsList = resource_search_paths;
+                    });
+            }
+            catch (EntryPointNotFoundException entryPointNotFoundException)
+            {
+                throw new InvalidOperationException(SR.ComponentDependencyResolver_FailedToLoadHostpolicy, entryPointNotFoundException);
+            }
+            catch (DllNotFoundException dllNotFoundException)
+            {
+                throw new InvalidOperationException(SR.ComponentDependencyResolver_FailedToLoadHostpolicy, dllNotFoundException);
+            }
+
+            if (returnCode != 0)
+            {
+                // Something went wrong - report a failure
+                // TODO - once we have the ability to capture detailed error messages from the hostpolicy
+                // use it here to make the exception more actionable.
+                throw new InvalidOperationException(SR.Format(SR.ComponentDependencyResolver_FailedToResolveDependencies, componentAssemblyPath, returnCode));
```

Comment:
This one might be the incorrect exception also....  It is not obvious why this would happen.

Reply:
This is thrown when anything goes wrong with processing the `.deps.json`. With the future error redirection work, we will get exact error details, so the exception message will be much better. We could also change exception types based on the various error codes returned by the native code.

Typical example where this will fire:
* The path to the component assembly is wrong (the file doesn't exist)
* Invalid JSON in the `.deps.json`
* Semantical error in the `.deps.json` - although there aren't that many cases where this would happen. The `hostpolicy` is pretty resilient to invalid inputs. (The one I know of is if there are duplicate entries in the `.deps.json` in which case `hostpolicy` doesn't know which one to pick)